engines/wintermute/base/gfx/opengl/x/modelx.cpp

```diff
@@ -0,0 +1,856 @@
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::loadFromFile(const Common::String &filename, ModelX *parentModel) {
+	cleanup(false);
+
+	uint32 fileSize = 0;
+	byte *buffer = BaseFileManager::getEngineInstance()->getEngineInstance()->readWholeFile(filename, &fileSize);
+
+	byte *dataFormatBlock = buffer + 8;
+
+	bool textMode = strcmp((char *)dataFormatBlock, "txt");
+
+	if (strcmp((char *)dataFormatBlock, "bzip") == 0 || strcmp((char *)dataFormatBlock, "tzip") == 0) {
+		warning("ModelX::loadFromFile compressed .X files are not supported yet");
+	}
+
+	XFileLexer lexer(buffer + 16, buffer + fileSize, textMode);
+
+	bool res = true;
+
+	_parentModel = parentModel;
+	_rootFrame = new FrameNode(_gameRef);
+	res = _rootFrame->loadFromXAsRoot(filename, lexer, this);
+	setFilename(filename.c_str());
+
+	for (int i = 0; i < X_NUM_ANIMATION_CHANNELS; ++i) {
+		_channels[i] = new AnimationChannel(_gameRef, this);
+	}
+
+	if (res) {
+		findBones(false, parentModel);
+	}
+
+	return res;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::loadAnimationSet(XFileLexer &lexer, const Common::String &filename) {
+	bool res = true;
+
+	AnimationSet *animSet = new AnimationSet(_gameRef, this);
+
+	if (animSet->loadFromX(lexer, filename)) {
+		_animationSets.add(animSet);
+	} else {
+		delete animSet;
+		res = false;
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::LoadAnimation(const Common::String &filename, AnimationSet *parentAnimSet) {
+	// not sure if we need this here (not completely implemented anyways and also not called)
+	// are there animation objects in .X outside of an animation set?
+
+	// if no parent anim set is specified, create one
+	bool newAnimSet = false;
+	if (parentAnimSet == nullptr) {
+		parentAnimSet = new AnimationSet(_gameRef, this);
+
+		parentAnimSet->setName(PathUtil::getFileName(filename).c_str());
+		newAnimSet = true;
+	}
+
+	// create the new object
+	Animation *Anim = new Animation(_gameRef);
+
+	parentAnimSet->addAnimation(Anim);
+
+	if (newAnimSet) {
+		_animationSets.add(parentAnimSet);
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::findBones(bool animOnly, ModelX *parentModel) {
+	FrameNode *rootFrame;
+	if (parentModel == nullptr)
+		rootFrame = _rootFrame;
+	else
+		rootFrame = parentModel->getRootFrame();
+
+	if (rootFrame && !animOnly) {
+		_rootFrame->findBones(rootFrame);
+	}
+
+	for (uint32 i = 0; i < _animationSets.size(); i++) {
+		_animationSets[i]->findBones(rootFrame);
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::loadName(BaseNamedObject *obj) {
+	return loadName(&obj->_name);
+}
+
+//////////////////////////////////////////////////////////////////////////
+bool ModelX::loadName(char **targetStr) {
+	uint32 nameLen;

```

Comment:
There's something left to do here.

Reply:
Actually this is irrelevant now, like a lot of other stuff. I will just remove it.